### DISCUSSION TITLE
Pyright Fails without variable assignment with SQLAlchemy Select type

### DISCUSSION THREAD

**Discussion by ManiMozaffar** [sqlalchemy/sqlalchemy#11476] [Usage Questions] [answered]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

The function signature is saying the return type is select of string. but i'm selecting Literal[1]. so type checker must always complain. but it doesn't always! It is very weird, when I try to assign it to variable and return it, type checker complain. When i just return it, type checker doesn't complain.
It makes more sense for me to post this in pyright github issue page, but I tried to reproduce it by combination of tuple, generic, and type var, however I couldn't get reproduce it without using sqlalchemy.

![Screenshot 2024-06-11 at 18 16 43](https://github.com/sqlalchemy/sqlalchemy/assets/86913668/f0be67d5-eb65-455c-baef-fea5973e8081)

### To Reproduce

```python
from typing import Any, Generic, TypeVar

import sqlalchemy as sa


def correct() -> sa.Select[tuple[str]]:
    q = sa.select(sa.literal(1))
    return q


def wrong() -> sa.Select[tuple[str]]:
    return sa.select(sa.literal(1))
```

### Error

```
Expression of type "Select[Tuple[int]]" is incompatible with return type "Select[tuple[str]]"
  "Select[Tuple[int]]" is incompatible with "Select[tuple[str]]"
    Type parameter "_TP@Select" is invariant, but "Tuple[int]" is not the same as "tuple[str]"Pylance[reportReturnType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportReturnType)
```

### Versions

- OS: MacOS
- Python: 3.11.1
- SQLAlchemy: 2.0.27
- Type checker (eg: mypy 0.991, pyright 1.1.290, etc): v1.1.350

**Comment by CaselIT** [comment]
Hi,

This seems a pyright issue. I don't see what sqlalchemy could do here to make things different.

**Comment by CaselIT** [comment, marked as answer]
Seems to be as designed by pyright: https://github.com/microsoft/pyright/issues/8120

**Reply by CaselIT** [reply]
I'm not sure if it's possible to change the overload in a way that makes this case work correctly.
IIRC the python typing do not support this use case yet, since we would need something like
```py
select(*entities: _ColumnsClauseArgument[Unpack[T]]) -> Select[Unpack[T]]
```
that was not supported the last time I checked
